### PR TITLE
Heterogen table

### PR DESCRIPTION
**Pull request by KulikovNikita**

### CODE REVIEW

**Comment by Alexsandruss** [comment]
No description provided.
Commit messages are not informative.

File: cpp/oneapi/dal/detail/array_impl.hpp [hunk 1]
```diff
@@ -15,7 +15,7 @@
 *******************************************************************************/
 
 #pragma once
-
+#include <iostream>
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/detail/array_impl.hpp:18]: please remove debug includes

File: cpp/oneapi/dal/table/detail/table_builder.hpp [hunk 2]
```diff
@@ -126,9 +127,80 @@ class ONEDAL_EXPORT homogen_table_builder : public table_builder {
     }
 };
 
+/*class ONEDAL_EXPORT heterogen_table_builder : public table_builder {
+public:
+    heterogen_table_builder();
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/table/detail/table_builder.hpp:132]: What we are going todo with this code?

**Review by samir-nasibli — COMMENTED** [review]
partially reviewed. Will complete it later

File: cpp/oneapi/dal/chunked_array.hpp [hunk 3]
```diff
@@ -0,0 +1,317 @@
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/chunked_array.hpp:2]: Year.

File: cpp/oneapi/dal/detail/chunked_array_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,219 @@
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/detail/chunked_array_impl.hpp:2]: year

File: cpp/oneapi/dal/table/backend/convert/common.cpp [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        auto row_size = detail::check_mul_overflow(raw_size, col_count);
+
+        offset = detail::check_sum_overflow(offset, row_size);
+        result_ptr[row] = offset;
+    }
+
+    return result;
+}
+
+} // namespace oneapi::dal::backend
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/backend/convert/common.cpp:171]: EOF

File: cpp/oneapi/dal/table/backend/convert/common.hpp [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                                                   std::int64_t count);
+
+dal::array<std::int64_t> find_unique_chunk_offsets(const dal::array<std::int64_t>& indices, //
+                                                   const dal::array<data_type>& inp,
+                                                   const dal::array<data_type>& out);
+
+dal::array<std::int64_t> find_sets_of_unique_pairs(const dal::array<data_type>& inp,
+                                                   const dal::array<data_type>& out);
+
+} // namespace oneapi::dal::backend
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/backend/convert/common.hpp:93]: EOF

File: cpp/oneapi/dal/table/backend/convert/common_convert.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    return compute_input_offsets(input_shape, input_types.get_data());
+}
+
+template dal::array<dal::byte_t*> compute_pointers<true>(const dal::array<dal::byte_t>&,
+                                                         const dal::array<std::int64_t>&);
+
+template dal::array<const dal::byte_t*> compute_pointers<false>(const dal::array<dal::byte_t>&,
+                                                                const dal::array<std::int64_t>&);
+
+} // namespace oneapi::dal::backend
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/backend/convert/common_convert.cpp:109]: EOF

File: cpp/oneapi/dal/table/backend/convert/copy_convert_dpc.cpp [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+                                  chunk_shape,
+                                  { last_event });
+
+        first = last;
+    }
+
+    return last_event;
+}
+
+} // namespace oneapi::dal::backend
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/backend/convert/copy_convert_dpc.cpp:177]: EOF

File: cpp/oneapi/dal/test/chunked_array_serialization.cpp [hunk 9]
```diff
@@ -0,0 +1,123 @@
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/test/chunked_array_serialization.cpp:2]: Year

**Comment by samir-nasibli** [comment]
@KulikovNikita many thanks for feature impl.
I see some failures public CI, eg docbuild, and also there are some of them on intelci.
Please address them on current PR or just disable this test and let's work on fixes on separate branches

File: cpp/oneapi/dal/table/heterogen.cpp [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    auto& impl = detail::cast_impl<detail::heterogen_table_iface>(*this);
+    ONEDAL_ASSERT(dt == this->get_metadata().get_data_type(column));
+    impl.set_column(column, dt, std::move(arr));
+}
+
+std::pair<data_type, detail::chunked_array_base> heterogen_table::get_column_impl(
+    std::int64_t column) const {
+    data_type dtype = this->get_metadata().get_data_type(column);
+    auto array = detail::cast_impl<const detail::heterogen_table_iface>(*this).get_column(column);
+    return std::make_pair<data_type, detail::chunked_array_base>(std::move(dtype),
```
Comment (ethanglaser) [inline cpp/oneapi/dal/table/heterogen.cpp:58]: To address codefactor, could alternatively do:
return std::make_pair(std::move(dtype), std::move(array));
or
return std::pair<data_type, detail::chunked_array_base>(std::move(dtype), std::move(array));

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/ee45ae77-05aa-f140-a4f0-a4bf010d0e2e

**Review by maria-Petrova — APPROVED** [review]

**Comment by ethanglaser** [comment]
Please address docbuild error: docs/api/data-management/array:79: WARNING: cpp:identifier reference target not found: impl_t

**Comment by ethanglaser** [comment]
> http://intel-ci.intel.com/ee45b116-af9e-f14c-b9fb-a4bf010d0e2e

Please rebase and re-run

File: cpp/oneapi/dal/chunked_array.hpp [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+
+    /// @brief       Move constrtuctor
+    /// @param array Source array
+    chunked_array(chunked_array&& array) : chunked_array_base{ std::forward<base_t>(array) } {}
+
+    /// @brief       Zero copy construct `chunked_array`
+    /// @param array Source array
+    chunked_array(const chunked_array& array) : chunked_array_base{ array } {}
+
+    /// @brief             Coonstructs an empty `chunked_array`
```
Comment (ethanglaser) [inline cpp/oneapi/dal/chunked_array.hpp:53]:
```suggestion
    /// @brief             Constructs an empty `chunked_array`
```

File: cpp/oneapi/dal/chunked_array.hpp [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+    friend class detail::array_impl;
+
+    friend detail::pimpl_accessor;
+
+    friend detail::serialization_accessor;
+
+public:
+    using data_t = T;
+
+    /// @brief       Move constrtuctor
```
Comment (ethanglaser) [inline cpp/oneapi/dal/chunked_array.hpp:45]:
```suggestion
    /// @brief       Move constructor
```

File: cpp/oneapi/dal/table/backend/register_serializable.cpp [hunk 13]
```diff
@@ -16,13 +16,16 @@
 
 #include "oneapi/dal/table/backend/empty_table_impl.hpp"
 #include "oneapi/dal/table/backend/homogen_table_impl.hpp"
+#include "oneapi/dal/table/backend/heterogen_table_impl.hpp"
 #include "oneapi/dal/table/backend/csr_table_impl.hpp"
 
 using oneapi::dal::backend::empty_table_impl;
 using oneapi::dal::backend::homogen_table_impl;
+//using oneapi::dal::backend::heterogen_table_impl;
```
Comment (ethanglaser) [inline cpp/oneapi/dal/table/backend/register_serializable.cpp:24]: can remove line completely if not using, same with line 29

Comment (KulikovNikita) [reply cpp/oneapi/dal/table/backend/register_serializable.cpp:24]: I'll add FIXME comment instead

**Comment by KulikovNikita** [comment]
http://intel-ci.intel.com/ee45d1f3-62be-f12f-94b4-a4bf010d0e2e

File: cpp/oneapi/dal/table/backend/register_serializable.cpp (hunk 13 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/table/backend/register_serializable.cpp]: fixed

**Review by samir-nasibli — APPROVED** [review]
Looks good to me. Let's wait green CI and merge.
I hope we will address all todos/test fails in separate PRs

File: docs/source/conf.py [hunk 14]
```diff
@@ -319,8 +319,9 @@
     ('cpp:identifier', 'sparse_indexing::zero_based'),
     ('cpp:identifier', 'sparse_indexing::one_based'),
     # array
-    ('cpp:identifier', 'array'),
     ('cpp:identifier', 'T'),
+    ('cpp:identifier', 'array'),
+    ('cpp.identifier', 'impl_t'),
```
Comment (ethanglaser) [inline docs/source/conf.py:324]: I think an update here requires update in docs (leads to docbuild error) so either update docs as well or remove this for now

**Review by ethanglaser — APPROVED** [review]
CI+code LGTM, docbuild can be addressed in follow-up.

**Pull request merged**